### PR TITLE
Added gNMI server 

### PR DESCRIPTION
**Pull request by renukamanavalan**
**To install:**
    Built LoM package for SONiC 
       e.g. remanava@remanava-dev-vm:~/sonic/fork/202205-sonic-buildimage$ ./src/sonic-device-health/LoM-Install/sonic/make-install-archive.sh 
    scp target/LoM-Install.bsx to switch.
    In the switch run, "bash ./LoM-Install.bsx -i"

**Test in SONiC switch:**
1, systemctl status device-health
2. sudo netstat -lntp
    tcp6       0      0 :::50059                :::*                    LISTEN      1272734/LoMgNMIServ 

Took gnmi_cli from "docker cp docker-device-health:/usr/bin/gnmi_cli ."
Ran
 
/home/admin/tools/gnmi_cli -client_types=gnmi -a 127.0.0.1:50059 -t EVENTS -logtostderr -insecure -v 7 -qt s -q 'all[heartbeat=5]' -streaming_type ON_CHANGE

```
Aggregate file coverages : 
_____________________________________________________________________________________
lom/src/engine/engine/context.go (99.3%))
lom/src/engine/engine/engine.go (83.3%))
lom/src/engine/engine/sequenceHandler.go (97.2%))
lom/src/engine/engine/serverReqHandler.go (100.0%))
lom/src/lib/lomcommon/config.go (97.0%))
lom/src/lib/lomcommon/helper.go (92.5%))
lom/src/lib/lomipc/client_transport.go (100.0%))
lom/src/lib/lomipc/json_transport.go (94.9%))
lom/src/lib/lomipc/server_transport.go (96.3%))
lom/src/lib/lomscripted/lom_api_ids.go (100.0%))
lom/src/lib/lomscripted/lom_script.go (99.2%))
lom/src/lib/lomscripted/lom_script_assist.go (100.0%))
lom/src/lib/lomtelemetry/channel.go (88.9%))
lom/src/lib/lomtelemetry/client.go (98.8%))
lom/src/lib/lomtelemetry/zmq_channel.go (95.4%))
lom/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go (97.5%))
lom/src/plugins/plugins_common/plugins_common.go (89.7%))
lom/src/plugins/plugins_common/plugins_helper.go (100.0%))
lom/src/plugins/plugins_files/sonic/client/dbclient/counter_repository.go (97.1%))
lom/src/plugins/plugins_files/sonic/client/dbclient/redis_provider.go (95.5%))
lom/src/plugins/plugins_files/sonic/plugin/linkcrc/link_crc.go (96.6%))
lom/src/gnmi/gnmi_server/server/basicAuth.go (100.0%))
lom/src/gnmi/gnmi_server/server/clientCertAuth.go (100.0%))
lom/src/gnmi/gnmi_server/server/client_subscribe.go (100.0%))
lom/src/gnmi/gnmi_server/server/jwtAuth.go (100.0%))
lom/src/gnmi/gnmi_server/server/pamAuth.go (95.3%))
lom/src/gnmi/gnmi_server/server/server.go (99.1%))
lom/src/gnmi/lom_data_clients/client.go (100.0%))
lom/src/gnmi/lom_data_clients/lom_data_client.go (100.0%))
```
_____________________________________________________________________________________

### CODE REVIEW

File: lom/src/gnmi/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# SONiC-telemetry
```
Comment (renukamanavalan) [inline lom/src/gnmi/README.md]: Drop it

File: lom/src/gnmi/gnmi_server/gnmi_server.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
+package main
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/gnmi_server.go:1]: Copy of SONiC telemetry -- Skim through

File: lom/src/gnmi/gnmi_server/server/basicAuth.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package gnmi
+
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/basicAuth.go:2]: *Auth.go -- Complete copy of SONiC telemetry -- Read it for knowledge only.

File: lom/src/gnmi/gnmi_server/server/client_subscribe.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
+package gnmi
+
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/client_subscribe.go:2]: 90% copy of SONiC. Only diff -- create Lom Client -- Mostly skim through

File: lom/src/gnmi/gnmi_server/server/client_subscribe.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    switch mode {
+    case gnmipb.SubscriptionList_STREAM:
+    default:
+        return grpc.Errorf(codes.InvalidArgument, "Unkown subscription mode: %q", query)
+    }
+
+    cmn.LogInfo("mode=%v, prefix=%q, target=%q", mode, prefix, target)
+
+    if (target == "COUNTERS") || (target == "EVENTS") {
+        dc, err = ldc.NewLoMDataClient(paths[0], prefix)
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/client_subscribe.go:134]: Only code change from SONiC here.

File: lom/src/gnmi/gnmi_server/server/jwtAuth.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
+package gnmi
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/jwtAuth.go:1]: Skip review - Copy of SONiC

File: lom/src/gnmi/gnmi_server/server/pamAuth.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
+package gnmi
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/pamAuth.go:1]: Skip review - Copy of SONiC

File: lom/src/gnmi/gnmi_server/server/server.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
+package gnmi
```
Comment (renukamanavalan) [inline lom/src/gnmi/gnmi_server/server/server.go:1]: 90% copy of SONiC - Just disabled Get and restricted capabilities -- Read for knowledge only

File: lom/src/gnmi/lom_data_clients/lom_data_client.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
+// Client to return counters & events
+package client
```
Comment (renukamanavalan) [inline lom/src/gnmi/lom_data_clients/lom_data_client.go:2]: 60% copy/paste

File: lom/src/gnmi/lom_data_clients/lom_data_client.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                c.onChg, _ = strconv.ParseBool(v)
+            }
+        }
+    }
+
+    cmn.LogDebug("%s Subscribe params (UpdFreq=%d PQ-Max=%d OnChange=%v)",
+        c.chTypeStr, c.updFreq, c.pq_max, c.onChg)
+
+    /* Init subscriber with cache use and defined time out */
+    if c.chData, c.chClose, err = tele.GetSubChannel(c.chType,
```
Comment (renukamanavalan) [inline lom/src/gnmi/lom_data_clients/lom_data_client.go:135]: This is the only LoM code

File: lom/src/gnmi/patches/0001-Updated-to-filter-and-write-to-file.patch [hunk 11]
```diff
@@ -0,0 +1,178 @@
+From c4660f5f1a5c38902c374f002174848aff2f1739 Mon Sep 17 00:00:00 2001
```
Comment (renukamanavalan) [inline lom/src/gnmi/patches/0001-Updated-to-filter-and-write-to-file.patch:1]: All patches are as of from SONiC

All packages are at the same commit as in SONiC

File: lom/src/gnmi/proto/lom_gnoi_jwt.proto [hunk 12]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
```
Comment (renukamanavalan) [inline lom/src/gnmi/proto/lom_gnoi_jwt.proto:1]: copy/paste

File: lom/src/gnmi/tools/dummy.go [hunk 13]
```diff
@@ -0,0 +1,3 @@
+package tools
```
Comment (renukamanavalan) [inline lom/src/gnmi/tools/dummy.go:1]: Do we need this?

File: lom/src/gnmi/utils/context.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
+package lom_utils
+
```
Comment (renukamanavalan) [inline lom/src/gnmi/utils/context.go:2]: Do we need?

File: lom/src/gnmi/utils/shareMem.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
+package lom_utils
```
Comment (renukamanavalan) [inline lom/src/gnmi/utils/shareMem.go:1]: Do we need ?

File: lom/src/gnmi/README.md (hunk 1 above)
Comment (renukamanavalan) [reply lom/src/gnmi/README.md]: Done

File: lom/src/gnmi/tools/dummy.go (hunk 13 above)
Comment (renukamanavalan) [reply lom/src/gnmi/tools/dummy.go:1]: Yes. We do.

We need openconfig tool gnmi_cli but as patched. So we need to import, patch and then build. Hence include in .go file to auto import.

File: lom/src/gnmi/utils/shareMem.go (hunk 15 above)
Comment (renukamanavalan) [reply lom/src/gnmi/utils/shareMem.go:1]: Yes. We need this, later when we extend to gNMI Get & Set. Eventually when we start using DBus,  we will need it then too.

File: lom/src/gnmi/utils/context.go (hunk 14 above)
Comment (renukamanavalan) [reply lom/src/gnmi/utils/context.go:2]: Yes, we need context.

remanava@remanava-dev-vm:~/lom/gnmi/lom/src/gnmi$ git grep GetContext
gnmi_server/server/basicAuth.go:    rc, ctx := lom_utils.GetContext(ctx)
gnmi_server/server/clientCertAuth.go:    rc, ctx := lom_utils.GetContext(ctx)
gnmi_server/server/jwtAuth.go:    rc, ctx := lom_utils.GetContext(ctx)
gnmi_server/server/server.go:    rc, ctx := lom_utils.GetContext(ctx)

File: LoM-Install/decompress.j2 [hunk 16]
```diff
@@ -1,41 +1,83 @@
...
-VERSION={{ LoMVersion }}
-if  [[ $# == 1 && "$1" == "-v" ]]; then
-    echo ${VERSION}
-    exit 0
-fi
+function usage()
+{
+    echo -e "\
+        -i - Does install or upgrade \n\
+        -f - Force a clean up of backup for upgrade if needed \n\
```
Comment (GouthamBanala) [inline LoM-Install/decompress.j2]: -f is not used right ?

File: lom/Makefile [hunk 17]
```diff
@@ -6,17 +6,24 @@ USER = $(shell id -un)
...
+export GOPATH=/tmp/go
+endif
+
+VENDOR_DIR := vendor
 BUILD_DIR := build
 LIB_DIR := $(BUILD_DIR)/lib
 BIN_DIR := $(BUILD_DIR)/bin
 TEST_DIR := $(BUILD_DIR)/test
 INTEGRATION_TEST_DIR := integration_test
+GNMI_DIR := src/gnmi
```
Comment (GouthamBanala) [inline lom/Makefile:24]: for my knowledge, gnmi server need libzmq  library too right ?

File: lom/Makefile [hunk 18]
```diff
@@ -190,9 +235,10 @@ $(ENGINE_TEST): $(SRC_FILES) $(TEST_FILES)
 
 $(PLUGIN_MANAGER_TEST): $(SRC_FILES) $(TEST_FILES)
 	mkdir -p $(TEST_DIR)
-	${GO} test -p 1 $(TEST_FLAG) -coverprofile=$(TEST_DIR)/PluginManagerCoverprofile.out \
+	${GO} test $(TEST_FLAG) -coverprofile=$(TEST_DIR)/PluginManagerCoverprofile.out \
```
Comment (GouthamBanala) [inline lom/Makefile:238]: -p 1 might be needed.Intention is to run the tests sequentially rather than parallely

**Review by GouthamBanala — COMMENTED** [review]
minor comemnts in Makefile.

File: LoM-Install/decompress.j2 (hunk 16 above)
Comment (renukamanavalan) [reply LoM-Install/decompress.j2]: dropped. "-c" should do the job.

File: lom/Makefile (hunk 17 above)
Comment (renukamanavalan) [reply lom/Makefile:24]: Every one who uses lib/libtelemetry -- needs it

File: lom/Makefile (hunk 18 above)
Comment (renukamanavalan) [reply lom/Makefile:238]: It is there.

LIB_TEST_FLAGS_NOINLINE = -gcflags=all=-l -p 1

**Review by GouthamBanala — APPROVED** [review]

**Review by prithviraghavtm — APPROVED** [review]
Changes look good to me..

**Pull request merged**